src/libraries/System.Private.CoreLib/src/System/CharEnumerator.cs

```diff
@@ -35,22 +35,27 @@ public object Clone()
 
         public bool MoveNext()
         {
-            if (_index < (_str!.Length - 1))
+            int index = _index + 1;
+            string s = _str!;
+
+            if ((uint)index < (uint)s.Length)
             {
-                _index++;
-                _currentElement = _str[_index];
+                _currentElement = s[index];
+                _index = index;
                 return true;
             }
-            else
-                _index = _str.Length;
+
+            _index = s.Length;
             return false;
         }
 
         public void Dispose()
         {
             if (_str != null)
+            {
                 _index = _str.Length;
```

Comment:
```suggestion
```
This should not be needed. All methods are going hit null reference exceptions once `_str` is set to null, irespective of what `_index` is set to.

Reply:
Removing it caused test failures for accessing Current after MoveNext returned false, as in that case Current would end up returning the last element again rather than throwing.  I put it back for now.